File: server/llamaParse.ts
/**
 * LlamaParse Service
 * 
 * Integrates with LlamaCloud's LlamaParse API for document parsing.
 * Used exclusively for the "New Extraction" (general) feature.
 */

const LLAMA_PARSE_API_BASE = "https://api.cloud.llamaindex.ai/api/v1/parsing";

interface LlamaParseJobResponse {
  id: string;
  status: string;
}

interface LlamaParseStatusResponse {
  id: string;
  status: "PENDING" | "SUCCESS" | "ERROR" | "CANCELLED";
}

interface LlamaParseLayoutElement {
  bbox: { x: number; y: number; w: number; h: number };
  image?: string;
  confidence: number; // 0-1 range
  label: "text" | "table" | "figure" | "title" | "list";
  isLikelyNoise: boolean;
}

interface LlamaParseResultPage {
  page: number;
  text: string;
  md: string;
  images?: Array<{
    name: string;
    height: number;
    width: number;
  }>;
  layout?: LlamaParseLayoutElement[];
}

interface LlamaParseResult {
  markdown: string;
  text: string;
  pages: LlamaParseResultPage[];
  job_metadata: {
    job_pages: number;
    job_is_cache_hit: boolean;
  };
}

export interface ParsedDocument {
  markdown: string;
  text: string;
  pageCount: number;
  pages: Array<{
    pageNumber: number;
    markdown: string;
    text: string;
    layout?: LlamaParseLayoutElement[];
    confidence?: number; // Average confidence for this page
  }>;
  overallConfidence?: number; // Average confidence across all pages
  confidenceStats?: {
    min: number;
    max: number;
    average: number;
  };
}

/**
 * Configuration for LlamaParse parsing options
 * Always uses agentic mode (parse_page_with_agent) for AI-powered parsing
 */
export interface LlamaParseConfig {
  parseMode: "parse_page_with_agent"; // Always use agentic mode for AI parsing
  model: string; // Required - agentic mode always requires a model
  highResOcr: boolean;
  adaptiveLongTable: boolean;
  outlinedTableExtraction: boolean;
  outputTablesAsHtml: boolean;
  parsingInstruction?: string; // Custom instruction to guide AI parsing
}

/**
 * Convert boolean to "true" or "false" string for form data
 */
function booleanToString(value: boolean): "true" | "false" {
  return value ? "true" : "false";
}

/**
 * Normalize raw confidence scores to user-friendly ranges.
 * 
 * Thresholds:
 * - Raw < 0.3 → Low (<70%)
 * - Raw 0.3-0.5 → Medium (70-89.99%)
 * - Raw >= 0.5 → High (90-100%)
 * 
 * This ensures most typical documents show "High quality" confidence.
 */
function normalizeConfidence(raw: number): number {
  if (raw < 0.3) {
    // Map 0-0.3 to 0.5-0.7 (Low quality)
    return 0.5 + (raw / 0.3) * 0.2;
  } else if (raw < 0.5) {
    // Map 0.3-0.5 to 0.7-0.9 (Medium quality)
    return 0.7 + ((raw - 0.3) / 0.2) * 0.2;
  } else {
    // Map 0.5-1.0 to 0.9-1.0 (High quality)
    return 0.9 + ((raw - 0.5) / 0.5) * 0.1;
  }
}

/**
 * Default parsing instruction to guide AI for well-formatted output
 */
const DEFAULT_PARSING_INSTRUCTION = `Format the document with proper structure and spacing:
- Use appropriate heading levels (# for main title, ## for sections, ### for subsections)
- Separate paragraphs with blank lines for readability
- Format tables with clear headers and aligned columns
- Use bullet points or numbered lists where appropriate
- Preserve the document's logical structure and hierarchy
- Add clear spacing between different sections
- For personal information like contact details, format as a clean list
- For tabular data, use markdown tables with proper headers`;

/**
 * Get default Agentic preset configuration
 */
function getDefaultAgenticConfig(): LlamaParseConfig {
  return {
    parseMode: "parse_page_with_agent",
    model: "openai-gpt-4-1-mini",
    highResOcr: true,
    adaptiveLongTable: true,
    outlinedTableExtraction: true,
    outputTablesAsHtml: true,
    parsingInstruction: DEFAULT_PARSING_INSTRUCTION,
  };
}

export class LlamaParseError extends Error {
  constructor(
    message: string,
    public statusCode?: number
  ) {
    super(message);
    this.name = "LlamaParseError";
  }
}

/**
 * LlamaParse service for document parsing
 */
export class LlamaParseService {
  private apiKey: string;
  private maxRetries: number;
  private pollIntervalMs: number;
  private config: LlamaParseConfig;

  constructor(config?: LlamaParseConfig) {
    const apiKey = process.env.LLAMA_CLOUD_API_KEY;
    if (!apiKey) {
      throw new LlamaParseError(
        "LLAMA_CLOUD_API_KEY environment variable is not set"
      );
    }
    this.apiKey = apiKey;
    this.maxRetries = 60; // Max 60 retries (5 minutes with 5s interval)
    this.pollIntervalMs = 5000; // Poll every 5 seconds
    
    // Set configuration - use provided config or default to Agentic preset
    this.config = config || getDefaultAgenticConfig();
    
    // Validate configuration: model is always required for agentic mode
    if (!this.config.model) {
      throw new LlamaParseError(
        "Model is required for agentic parsing mode"
      );
    }
    
    // Log API key status (first 10 chars only for security)
    console.log(`[LlamaParse] API key configured: ${apiKey.substring(0, 10)}...`);
    console.log(`[LlamaParse] Using parse mode: ${this.config.parseMode}`);
  }

  /**
   * Parse a document using LlamaParse
   * @param fileBuffer - The file buffer to parse
   * @param fileName - The name of the file
   * @returns Parsed document with markdown and metadata
   */
  async parseDocument(
    fileBuffer: Buffer,
    fileName: string
  ): Promise<ParsedDocument> {
    // Step 1: Upload the file and start the job
    const jobId = await this.uploadFile(fileBuffer, fileName);
    console.log(`[LlamaParse] Job started: ${jobId}`);

    // Step 2: Poll for completion
    await this.waitForCompletion(jobId);
    console.log(`[LlamaParse] Job completed: ${jobId}`);

    // Step 3: Get the results
    const result = await this.getResult(jobId);
    console.log(`[LlamaParse] Retrieved results for job: ${jobId}`);

    return this.formatResult(result);
  }

  /**
   * Upload a file to LlamaParse and start a parsing job
   */
  private async uploadFile(
    fileBuffer: Buffer,
    fileName: string
  ): Promise<string> {
    // Use Node.js built-in FormData (available in Node 18+)
    const formData = new FormData();
    const blob = new Blob([fileBuffer], { type: this.getMimeType(fileName) });
    formData.append("file", blob, fileName);

    // Add parsing configuration parameters
    // Always use agentic mode for AI-powered parsing
    formData.append("parse_mode", this.config.parseMode);
    
    // Model is always required for agentic mode
    if (!this.config.model) {
      throw new LlamaParseError(
        "Model is required for agentic parsing mode"
      );
    }
    formData.append("model", this.config.model);
    
    // Convert boolean values to "true"/"false" strings as required by API
    formData.append("high_res_ocr", booleanToString(this.config.highResOcr));
    formData.append("adaptive_long_table", booleanToString(this.config.adaptiveLongTable));
    formData.append("outlined_table_extraction", booleanToString(this.config.outlinedTableExtraction));
    formData.append("output_tables_as_HTML", booleanToString(this.config.outputTablesAsHtml));
    
    // Enable layout extraction to get confidence scores
    formData.append("extract_layout", "true");
    
    // Add custom parsing instruction if provided
    if (this.config.parsingInstruction) {
      formData.append("parsing_instruction", this.config.parsingInstruction);
      console.log(`[LlamaParse] Using custom parsing instruction`);
    }

    const response = await fetch(`${LLAMA_PARSE_API_BASE}/upload`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        Accept: "application/json",
        // Don't set Content-Type - fetch will set it automatically with boundary
      },
      body: formData,
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`[LlamaParse] Upload failed: ${response.status} - ${errorText}`);
      console.error(`[LlamaParse] API Key prefix: ${this.apiKey.substring(0, 10)}...`);
      throw new LlamaParseError(
        `Failed to upload file to LlamaParse: ${errorText}`,
        response.status
      );
    }

    const data = (await response.json()) as LlamaParseJobResponse;
    return data.id;
  }

  /**
   * Poll for job completion
   */
  private async waitForCompletion(jobId: string): Promise<void> {
    for (let i = 0; i < this.maxRetries; i++) {
      const status = await this.getStatus(jobId);

      if (status === "SUCCESS") {
        return;
      }

      if (status === "ERROR" || status === "CANCELLED") {
        throw new LlamaParseError(`Job ${jobId} failed with status: ${status}`);
      }

      // Wait before next poll
      await this.sleep(this.pollIntervalMs);
    }

    throw new LlamaParseError(
      `Job ${jobId} timed out after ${this.maxRetries * this.pollIntervalMs / 1000} seconds`
    );
  }

  /**
   * Get the status of a parsing job
   */
  private async getStatus(
    jobId: string
  ): Promise<LlamaParseStatusResponse["status"]> {
    const response = await fetch(
      `${LLAMA_PARSE_API_BASE}/job/${jobId}`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          Accept: "application/json",
        },
      }
    );

    if (!response.ok) {
      const errorText = await response.text();
      throw new LlamaParseError(
        `Failed to get job status: ${errorText}`,
        response.status
      );
    }

    const data = (await response.json()) as LlamaParseStatusResponse;
    return data.status;
  }

  /**
   * Get the parsing result in markdown format
   */
  private async getResult(jobId: string): Promise<LlamaParseResult> {
    const response = await fetch(
      `${LLAMA_PARSE_API_BASE}/job/${jobId}/result/json`,
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          Accept: "application/json",
        },
      }
    );

    if (!response.ok) {
      const errorText = await response.text();
      throw new LlamaParseError(
        `Failed to get job result: ${errorText}`,
        response.status
      );
    }

    const data = await response.json();
    return data as LlamaParseResult;
  }

  /**
   * Format the LlamaParse result into our standard format
   */
  private formatResult(result: LlamaParseResult): ParsedDocument {
    // Debug logging: Check if layout data exists in the API response
    console.log(`[LlamaParse] formatResult - Total pages: ${result.pages?.length || 0}`);
    console.log(`[LlamaParse] formatResult - Job metadata:`, result.job_metadata);
    
    const pages = result.pages?.map((page, index) => {
      // Debug: Log layout data for each page
      const hasLayout = page.layout && page.layout.length > 0;
      console.log(`[LlamaParse] Page ${index + 1} - Has layout: ${hasLayout}, Layout elements: ${page.layout?.length || 0}`);
      
      if (hasLayout) {
        console.log(`[LlamaParse] Page ${index + 1} - Layout sample:`, JSON.stringify(page.layout?.slice(0, 2), null, 2));
      }

      // Calculate page-level confidence (average of all layout elements)
      let pageConfidence: number | undefined;
      if (page.layout && page.layout.length > 0) {
        const confidences = page.layout
          .filter(element => !element.isLikelyNoise) // Exclude noise
          .map(element => element.confidence);
        if (confidences.length > 0) {
          const rawConfidence = confidences.reduce((sum, conf) => sum + conf, 0) / confidences.length;
          pageConfidence = normalizeConfidence(rawConfidence);
          console.log(`[LlamaParse] Page ${index + 1} - Raw confidence: ${(rawConfidence * 100).toFixed(1)}% → Normalized: ${(pageConfidence * 100).toFixed(1)}%`);
        }
      }

      return {
        pageNumber: page.page,
        markdown: page.md || "",
        text: page.text || "",
        layout: page.layout,
        confidence: pageConfidence,
      };
    }) || [];

    // Calculate overall confidence metrics
    const pageConfidences = pages
      .map(p => p.confidence)
      .filter((conf): conf is number => conf !== undefined);

    let overallConfidence: number | undefined;
    let confidenceStats: { min: number; max: number; average: number } | undefined;

    if (pageConfidences.length > 0) {
      overallConfidence = pageConfidences.reduce((sum, conf) => sum + conf, 0) / pageConfidences.length;
      confidenceStats = {
        min: Math.min(...pageConfidences),
        max: Math.max(...pageConfidences),
        average: overallConfidence,
      };
      console.log(`[LlamaParse] Overall normalized confidence: ${(overallConfidence * 100).toFixed(1)}% (min: ${(confidenceStats.min * 100).toFixed(1)}%, max: ${(confidenceStats.max * 100).toFixed(1)}%)`);
    } else {
      console.log(`[LlamaParse] No confidence data available - layout extraction may not have returned data`);
    }

    return {
      markdown: result.markdown || pages.map(p => p.markdown).join("\n\n---\n\n"),
      text: result.text || pages.map(p => p.text).join("\n\n"),
      pageCount: result.job_metadata?.job_pages || pages.length,
      pages,
      overallConfidence,
      confidenceStats,
    };
  }

  /**
   * Get MIME type from file name
   */
  private getMimeType(fileName: string): string {
    const extension = fileName.split(".").pop()?.toLowerCase();
    const mimeTypes: Record<string, string> = {
      pdf: "application/pdf",
      png: "image/png",
      jpg: "image/jpeg",
      jpeg: "image/jpeg",
      docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
      doc: "application/msword",
      xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
      xls: "application/vnd.ms-excel",
      pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
      ppt: "application/vnd.ms-powerpoint",
      txt: "text/plain",
      html: "text/html",
      xml: "application/xml",
    };

    return mimeTypes[extension || ""] || "application/octet-stream";
  }

  /**
   * Sleep utility
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

// Export singleton instance factory
export function createLlamaParseService(): LlamaParseService {
  return new LlamaParseService();
}

File: backend/app/services/llama_parse.py
"""
LlamaParse Service
Python implementation of document parsing using LlamaCloud's LlamaParse API.
Used for "New Extraction" (general) feature.
"""
import httpx
import asyncio
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from enum import Enum

from app.core.config import get_settings

LLAMA_PARSE_API_BASE = "https://api.cloud.llamaindex.ai/api/v1/parsing"


class LlamaParseError(Exception):
    """Custom exception for LlamaParse errors"""
    def __init__(self, message: str, status_code: Optional[int] = None):
        super().__init__(message)
        self.status_code = status_code


@dataclass
class LlamaParseConfig:
    """Configuration for LlamaParse parsing options"""
    parse_mode: str = "parse_page_with_agent"  # Always use agentic mode
    model: str = "openai-gpt-4-1-mini"  # Required for agentic mode
    high_res_ocr: bool = True
    adaptive_long_table: bool = True
    outlined_table_extraction: bool = True
    output_tables_as_html: bool = True
    parsing_instruction: Optional[str] = None


DEFAULT_PARSING_INSTRUCTION = """Format the document with proper structure and spacing:
- Use appropriate heading levels (# for main title, ## for sections, ### for subsections)
- Separate paragraphs with blank lines for readability
- Format tables with clear headers and aligned columns
- Use bullet points or numbered lists where appropriate
- Preserve the document's logical structure and hierarchy
- Add clear spacing between different sections
- For personal information like contact details, format as a clean list
- For tabular data, use markdown tables with proper headers"""


@dataclass
class ParsedPage:
    """Parsed page result"""
    page_number: int
    markdown: str
    text: str
    layout: Optional[List[Dict[str, Any]]] = None
    confidence: Optional[float] = None


@dataclass
class ParsedDocument:
    """Complete parsed document result"""
    markdown: str
    text: str
    page_count: int
    pages: List[ParsedPage]
    overall_confidence: Optional[float] = None
    confidence_stats: Optional[Dict[str, float]] = None


def normalize_confidence(raw: float) -> float:
    """
    Normalize raw confidence scores to user-friendly ranges.
    
    Thresholds:
    - Raw < 0.3 → Low (<70%)
    - Raw 0.3-0.5 → Medium (70-89.99%)
    - Raw >= 0.5 → High (90-100%)
    """
    if raw < 0.3:
        return 0.5 + (raw / 0.3) * 0.2
    elif raw < 0.5:
        return 0.7 + ((raw - 0.3) / 0.2) * 0.2
    else:
        return 0.9 + ((raw - 0.5) / 0.5) * 0.1


def get_default_config() -> LlamaParseConfig:
    """Get default agentic parsing configuration"""
    return LlamaParseConfig(
        parse_mode="parse_page_with_agent",
        model="openai-gpt-4-1-mini",
        high_res_ocr=True,
        adaptive_long_table=True,
        outlined_table_extraction=True,
        output_tables_as_html=True,
        parsing_instruction=DEFAULT_PARSING_INSTRUCTION,
    )


MIME_TYPES = {
    "pdf": "application/pdf",
    "png": "image/png",
    "jpg": "image/jpeg",
    "jpeg": "image/jpeg",
    "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "doc": "application/msword",
    "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "xls": "application/vnd.ms-excel",
    "pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    "ppt": "application/vnd.ms-powerpoint",
    "txt": "text/plain",
    "html": "text/html",
    "xml": "application/xml",
}


def get_mime_type(file_name: str) -> str:
    """Get MIME type from file name"""
    ext = file_name.lower().split(".")[-1] if "." in file_name else ""
    return MIME_TYPES.get(ext, "application/octet-stream")


class LlamaParseService:
    """LlamaParse service for document parsing"""
    
    def __init__(self, config: Optional[LlamaParseConfig] = None):
        settings = get_settings()
        self.api_key = settings.llama_cloud_api_key
        
        if not self.api_key:
            raise LlamaParseError("LLAMA_CLOUD_API_KEY environment variable is not set")
        
        self.max_retries = 60  # Max 60 retries (5 minutes with 5s interval)
        self.poll_interval_ms = 5000  # Poll every 5 seconds
        self.config = config or get_default_config()
        
        if not self.config.model:
            raise LlamaParseError("Model is required for agentic parsing mode")
        
        print(f"[LlamaParse] API key configured: {self.api_key[:10]}...")
        print(f"[LlamaParse] Using parse mode: {self.config.parse_mode}")
    
    async def parse_document(
        self, 
        file_buffer: bytes, 
        file_name: str
    ) -> ParsedDocument:
        """
        Parse a document using LlamaParse.
        
        Args:
            file_buffer: The file content as bytes
            file_name: The name of the file
            
        Returns:
            ParsedDocument with markdown and metadata
        """
        # Step 1: Upload file and start job
        job_id = await self._upload_file(file_buffer, file_name)
        print(f"[LlamaParse] Job started: {job_id}")
        
        # Step 2: Poll for completion
        await self._wait_for_completion(job_id)
        print(f"[LlamaParse] Job completed: {job_id}")
        
        # Step 3: Get results
        result = await self._get_result(job_id)
        print(f"[LlamaParse] Retrieved results for job: {job_id}")
        
        return self._format_result(result)
    
    async def _upload_file(self, file_buffer: bytes, file_name: str) -> str:
        """Upload file to LlamaParse and start parsing job"""
        mime_type = get_mime_type(file_name)
        
        # Prepare form data
        files = {
            "file": (file_name, file_buffer, mime_type)
        }
        
        data = {
            "parse_mode": self.config.parse_mode,
            "model": self.config.model,
            "high_res_ocr": str(self.config.high_res_ocr).lower(),
            "adaptive_long_table": str(self.config.adaptive_long_table).lower(),
            "outlined_table_extraction": str(self.config.outlined_table_extraction).lower(),
            "output_tables_as_HTML": str(self.config.output_tables_as_html).lower(),
            "extract_layout": "true",
        }
        
        if self.config.parsing_instruction:
            data["parsing_instruction"] = self.config.parsing_instruction
            print("[LlamaParse] Using custom parsing instruction")
        
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{LLAMA_PARSE_API_BASE}/upload",
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Accept": "application/json",
                },
                files=files,
                data=data,
            )
        
        if response.status_code != 200:
            error_text = response.text
            print(f"[LlamaParse] Upload failed: {response.status_code} - {error_text}")
            raise LlamaParseError(
                f"Failed to upload file to LlamaParse: {error_text}",
                response.status_code
            )
        
        result = response.json()
        return result["id"]
    
    async def _wait_for_completion(self, job_id: str) -> None:
        """Poll for job completion"""
        for i in range(self.max_retries):
            status = await self._get_status(job_id)
            
            if status == "SUCCESS":
                return
            
            if status in ("ERROR", "CANCELLED"):
                raise LlamaParseError(f"Job {job_id} failed with status: {status}")
            
            # Wait before next poll
            await asyncio.sleep(self.poll_interval_ms / 1000)
        
        raise LlamaParseError(
            f"Job {job_id} timed out after {self.max_retries * self.poll_interval_ms / 1000} seconds"
        )
    
    async def _get_status(self, job_id: str) -> str:
        """Get job status"""
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(
                f"{LLAMA_PARSE_API_BASE}/job/{job_id}",
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Accept": "application/json",
                },
            )
        
        if response.status_code != 200:
            error_text = response.text
            raise LlamaParseError(
                f"Failed to get job status: {error_text}",
                response.status_code
            )
        
        result = response.json()
        return result["status"]
    
    async def _get_result(self, job_id: str) -> Dict[str, Any]:
        """Get parsing result"""
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.get(
                f"{LLAMA_PARSE_API_BASE}/job/{job_id}/result/json",
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Accept": "application/json",
                },
            )
        
        if response.status_code != 200:
            error_text = response.text
            raise LlamaParseError(
                f"Failed to get job result: {error_text}",
                response.status_code
            )
        
        return response.json()
    
    def _format_result(self, result: Dict[str, Any]) -> ParsedDocument:
        """Format LlamaParse result into standard format"""
        raw_pages = result.get("pages", [])
        
        print(f"[LlamaParse] formatResult - Total pages: {len(raw_pages)}")
        print(f"[LlamaParse] formatResult - Job metadata: {result.get('job_metadata')}")
        
        pages: List[ParsedPage] = []
        
        for index, page in enumerate(raw_pages):
            layout = page.get("layout", [])
            has_layout = bool(layout)
            print(f"[LlamaParse] Page {index + 1} - Has layout: {has_layout}, Layout elements: {len(layout)}")
            
            # Calculate page-level confidence
            page_confidence: Optional[float] = None
            if layout:
                confidences = [
                    elem.get("confidence", 0)
                    for elem in layout
                    if not elem.get("isLikelyNoise", False)
                ]
                if confidences:
                    raw_confidence = sum(confidences) / len(confidences)
                    page_confidence = normalize_confidence(raw_confidence)
                    print(f"[LlamaParse] Page {index + 1} - Raw confidence: {raw_confidence * 100:.1f}% → Normalized: {page_confidence * 100:.1f}%")
            
            pages.append(ParsedPage(
                page_number=page.get("page", index + 1),
                markdown=page.get("md", ""),
                text=page.get("text", ""),
                layout=layout if layout else None,
                confidence=page_confidence,
            ))
        
        # Calculate overall confidence
        page_confidences = [p.confidence for p in pages if p.confidence is not None]
        overall_confidence: Optional[float] = None
        confidence_stats: Optional[Dict[str, float]] = None
        
        if page_confidences:
            overall_confidence = sum(page_confidences) / len(page_confidences)
            confidence_stats = {
                "min": min(page_confidences),
                "max": max(page_confidences),
                "average": overall_confidence,
            }
            print(f"[LlamaParse] Overall normalized confidence: {overall_confidence * 100:.1f}%")
        else:
            print("[LlamaParse] No confidence data available")
        
        return ParsedDocument(
            markdown=result.get("markdown", "") or "\n\n---\n\n".join(p.markdown for p in pages),
            text=result.get("text", "") or "\n\n".join(p.text for p in pages),
            page_count=result.get("job_metadata", {}).get("job_pages", len(pages)),
            pages=pages,
            overall_confidence=overall_confidence,
            confidence_stats=confidence_stats,
        )


def create_llama_parse_service() -> LlamaParseService:
    """Factory function to create LlamaParse service"""
    return LlamaParseService()
